qiskit/providers/aer/noise/passes.py

```diff
@@ -0,0 +1,209 @@
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+"""
+Noise addition passes.
+"""
+from typing import Optional, Union, Sequence, Callable, List, Iterable
+
+import numpy as np
+
+from qiskit.circuit import Instruction, QuantumCircuit
+from qiskit.dagcircuit import DAGCircuit
+from qiskit.transpiler import TransformationPass
+from qiskit.transpiler.exceptions import TranspilerError
+from .errors import QuantumError
+from .errors.standard_errors import thermal_relaxation_error
+
+InstructionLike = Union[Instruction, QuantumError]
+
+
+class LocalNoisePass(TransformationPass):
+    """Transpiler pass to insert noise into a circuit.
+
+    The noise in this pass is defined by a noise function or callable with signature
+
+    .. code:: python
+
+            def fn(
+                inst: Instruction,
+                qubits: Optional[List[int]] = None
+            ) -> InstructionLike:
+
+    For every instance of one of the reference instructions in a circuit the
+    supplied function is called on that instruction and the returned noise
+    is added to the circuit. This noise can depend on properties of the
+    instruction it is called on (for example parameters or duration) to
+    allow inserting parameterized noise models.
+
+    Several methods for adding the constructed errors to circuits are supported
+    and can be set by using the ``method`` kwarg. The supported methods are
+
+    * ``"append"``: add the return of the callable after the instruction.
+    * ``"prepend"``: add the return of the callable before the instruction.
+    * ``"replace"``: replace the instruction with the return of the callable.
+
+    """
+
+    def __init__(
+            self,
+            fn: Callable[[Instruction, Sequence[int]], InstructionLike],
+            ops: Optional[Union[Instruction, Iterable[Instruction]]] = None,
+            method: str = 'append'
+    ):
+        """Initialize noise pass.
+
+        Args:
+            fn: noise function `fn(inst, qubits) -> InstructionLike`.
+            ops: Optional, single or list of instructions to apply the
+                                  noise function to. If None the noise function will be
+                                  applied to all instructions in the circuit.
+            method: method for inserting noise. Allow methods are
+                           'append', 'prepend', 'replace'.
+        Raises:
+            TranspilerError: if an invalid option is specified.
+        """
+        if method not in {"append", "prepend", "replace"}:
+            raise TranspilerError(
+                f'Invalid method: {method}, it must be "append", "prepend" or "replace"'
+            )
+        if isinstance(ops, str):
+            ops = [ops]
+        super().__init__()
+        self._fn = fn
+        self._ops = set(ops) if ops else {}
```

Comment:
If you use strings instead of instances / objects this doesn't match the type hints. You would need to do `instruction.name` to add to the set. Personally I think using specific instructions would be better than using strings so this would be something like:

`self._ops = set(type(op) if isinstance(op, Instruction) else op for op in ops)`

Reply:
5753075